docs/internal/3i.md

```diff
@@ -0,0 +1,149 @@
+
+## Grates
+
+### Overview
+
+Grates are cages that provide services to descendant cages by interposing on their system calls. They can execute trusted operations (e.g., syscall filtering, tracing, or in-memory file I/O) on behalf of other cages without requiring changes to the microvisor’s trusted computing base (TCB).
+
+From 3i’s perspective, a grate is both:
+
+- A consumer of system calls from descendant cages, and
+- A service provider exposing one or more of its own functions to some set of other cages.
+
+### Design Goals
+
+3i’s API is runtime-agnostic, allowing it to be adapted to different backends (e.g., Wasmtime, MPK, NaCl, or future runtimes) and also supports simultaneous use of different backends.
+
+Note that some caging technologies may need to do specific actions upon cage entry and/or exit.  Since 3i supports calling arbitrary code, one can use this to jump to cage technology specific routines to handle those aspects.  This means that things such as Wasmtime’s `VMContext` management, are implemented in the corresponding runtime layer (e.g., the Wasmtime adapter). These runtime hooks are exposed to 3i as C-ABI function pointers and registered in 3i’s global table, decoupling runtime details from 3i’s core logic.
+
+### Wasm Implementation
+
+Grate execution relies on a re-entry mechanism that bridges Wasm runtime contexts (e.g., Wasmtime instances) and 3i’s routing layer. 
+
+Each Wasm module in Wasmtime has its own `VMContext`, which encapsulates all runtime state required for that instance’s execution. When 3i performs a cross-module call (e.g., from _cage A_ to a _grate B_), the currently executing context belongs to _A_, but 3i needs to load the latest runtime state of _B_ before re-entering _B_'s Wasm code. 
+
+Another design difference in lind-wasm is that each thread within a cage is represented by a separate Wasmtime instance. These per-thread instances share the same host memory by importing a common memory object, which enables in-process shared memory between threads. Each thread instance, however, maintains its own call stack. Because of this design, when storing runtime contexts we must uniquely identify them by both cage ID (`cageid`) and thread ID (`tid`). Most system calls can execute using the main-thread instance (`threadid = 0`), since they do not depend on per-thread stack state. However, calls such as `fork()`, which must capture or replicate the current thread’s execution stack, require referring to the precise thread-specific instance. To handle concurrent requests, lind-wasm maintains a small instance pool for the main-thread context: instead of a single threadid = 0 instance, we pre-allocate around ten clones that share the same memory but have separate stacks. This allows multiple inbound grate requests to execute in parallel without blocking on a single Wasm context.
```

Comment:
i think the thread per instance distinction needs to move closer to the beginning, also some confusion in module vs instance on my end